src/compiler/checker.ts

```diff
@@ -22469,6 +22469,10 @@ namespace ts {
             }
         }
 
+        function isCircularMappedProperty(symbol: Symbol) {
+            return !!(getCheckFlags(symbol) & CheckFlags.Mapped && !(<MappedSymbol>symbol).type && findResolutionCycleStartIndex(symbol, TypeSystemPropertyName.Type) >= 0);

```

Comment:
For my own edification: it looks like we're maintaining a stack of symbols to be resolved and `findResolutionCycleStartIndex` walks backwards through the stack to see if the current symbol is already in it (as the type of a property?)?

Reply:
Yes, we maintain a stack of type system entities whose resolution is in-process. If we see the same entity twice, we know we have a circularity and, importantly, we know all of the entities involved in the cycle and can error on each.

Reply:
And, here we just use the stack to discover if we're about to set off a circularity error.